File: jeph2sworm-vscode/src/providers/anthropic-provider.ts
/**
 * Anthropic Provider — proxy to Claude 4 / Claude 3.5 Sonnet / Haiku models.
 */

import { BaseProvider, LLMMessage, LLMResponse, ProviderConfig } from './base-provider';

export class AnthropicProvider extends BaseProvider {
  readonly name = 'anthropic';
  readonly models = ['claude-sonnet-4-20250514', 'claude-3-5-sonnet-20241022', 'claude-3-5-haiku-20241022'];

  constructor(config: ProviderConfig) {
    super({ baseUrl: 'https://api.anthropic.com/v1', ...config });
  }

  async complete(messages: LLMMessage[], model?: string): Promise<LLMResponse> {
    const chosen = model ?? this.config.defaultModel ?? 'claude-sonnet-4-20250514';
    const system = messages.find((m) => m.role === 'system')?.content;
    const filtered = messages.filter((m) => m.role !== 'system');

    const res = await fetch(`${this.config.baseUrl}/messages`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'x-api-key': this.config.apiKey,
        'anthropic-version': '2023-06-01',
      },
      body: JSON.stringify({
        model: chosen,
        max_tokens: 8192,
        system,
        messages: filtered,
      }),
    });
    const data = await res.json();
    return {
      content: data.content?.[0]?.text ?? '',
      model: chosen,
      tokensUsed: (data.usage?.input_tokens ?? 0) + (data.usage?.output_tokens ?? 0),
      finishReason: data.stop_reason ?? 'unknown',
    };
  }

  bestModelFor(task: 'coding' | 'planning' | 'design' | 'quick'): string {
    switch (task) {
      case 'coding':
        return 'claude-sonnet-4-20250514';
      case 'planning':
        return 'claude-sonnet-4-20250514';
      case 'design':
        return 'claude-3-5-sonnet-20241022';
      case 'quick':
        return 'claude-3-5-haiku-20241022';
    }
  }
}

File: jeph2sworm-core/jeph2sworm/llm/anthropic_provider.py
"""Anthropic Provider - Claude 3.5 Sonnet, Claude 3 Opus, Claude 3 Haiku."""

from __future__ import annotations

from typing import List, Optional

import structlog

from jeph2sworm.llm.base_provider import BaseProvider, LLMMessage, LLMResponse

logger = structlog.get_logger()

MODELS = [
    "claude-sonnet-4-20250514",
    "claude-3-5-sonnet-20241022",
    "claude-3-opus-20240229",
    "claude-3-haiku-20240307",
]

TASK_BEST = {
    "coding": "claude-sonnet-4-20250514",
    "planning": "claude-3-opus-20240229",
    "reasoning": "claude-3-opus-20240229",
    "design": "claude-sonnet-4-20250514",
    "testing": "claude-3-haiku-20240307",
    "general": "claude-sonnet-4-20250514",
}


class AnthropicProvider(BaseProvider):
    provider_name = "anthropic"
    supports_function_calling = True
    supports_vision = True
    max_context_window = 200_000

    def available_models(self) -> List[str]:
        return MODELS

    def best_model_for(self, task: str) -> str:
        return TASK_BEST.get(task, "claude-sonnet-4-20250514")

    async def complete(
        self,
        messages: List[LLMMessage],
        model: str,
        temperature: float = 0.7,
        max_tokens: int = 4096,
        **kwargs,
    ) -> LLMResponse:
        import litellm

        response = await litellm.acompletion(
            model=f"anthropic/{model}" if not model.startswith("anthropic/") else model,
            messages=[m.model_dump() for m in messages],
            temperature=temperature,
            max_tokens=max_tokens,
            api_key=self.api_key,
            **kwargs,
        )

        choice = response.choices[0]
        usage = dict(response.usage) if response.usage else {}

        return LLMResponse(
            content=choice.message.content or "",
            model=model,
            provider=self.provider_name,
            usage=usage,
            finish_reason=choice.finish_reason or "stop",
        )
